### ISSUE TITLE
Dependencies view should deduplicate edges

### ISSUE THREAD

**Issue by chemicL** [openzipkin/zipkin#868]
Due to the fact, that the UI does not specify the `startTs` param calling the query-service, duplicate edges might be added to the graph, which results in an exception on the js side:

```
Uncaught Error: Graph already has edge 'serviceA->serviceB'
```

The problem is caused by running the dependencies job on different days.

In terms of the UI, assuming a range of more than one day is displayed, the edge counts could be added and one unique edge with the sum would exist in the graph.
It could also be implemented on the java side, probably in `CassandraDependencyStore` to be analogous to `AnormDependencyStore`, which does the sum operation and provides unique edges.

**Comment by yurishkuro** [comment]
I just ran into the same thing. 

```
    FutureUtil.toFuture(repository.getDependencies(startEpochDayMillis, endEpochDayMillis))
      .map { dependencies => dependencies.asScala
          .map(codec.decode(_))
          .map(thriftToDependencies(_).toDependencies)
          .flatMap(_.links)
      }
```

The last `flatMap` there joins all edges from runs on different days. I don't think we want to do that, we just want to pick the entry with the latest timestamp, and only use its links.

cc @adriancole

**Comment by codefromthecrypt** [comment]
sounds good. most important will be to update DependencyStoreSpec to use
the least amount of dummy data to reproduce this.

**Comment by yurishkuro** [comment]
This fixes it:

```
    FutureUtil.toFuture(repository.getDependencies(startEpochDayMillis, endEpochDayMillis))
      .map { dependencies =>
        dependencies.asScala
          .map(codec.decode(_))
          .map(thriftToDependencies(_).toDependencies)
          .maxBy(_.endTs).links
      }
```

**Comment by codefromthecrypt** [comment]
Thanks, yuri. If no-one else can integrate this into a PR, I can probably
do so before the end of the day.

**Comment by chemicL** [comment]
Not sure if that's a desired behaviour - assuming we select dependencies for two days, we should get an aggregation of all the spans in that period. That's why summing the counts would make more sense than selecting just the latest aggregation. Otherwise, if on day_1 there were 5 spans from A to B, and 3 on day_2, the above solution would show A -> B with count of 3, while there were total 8 spans.
Unless I'm mistaken the sum approach would also be in line with the `AnormDependencyStore` implementation.

**Comment by codefromthecrypt** [comment]
Thanks for the test case!

**Comment by yurishkuro** [comment]
I would prefer if the UI could simply display the time interval for which the aggregated results are shown. Adding multiple records together doesn't sound right to me - what if their startTs-endTs intervals overlap or are disjoint? The aggregate numbers wouldn't really make sense. At least if we're showing a single aggregate record we can guarantee that the counts are consistent.

Also, summing may work for counts, but once we start adding more measures, like p99, we won't be able to aggregate across multiple time intervals (maybe we'll need to go back to 'moments').

Incidentally, the current Spark job impl leaves startTs/endTs==0, so the maxBY above won't quite work.

```
    val dependencies: Dependencies = aggregates
      .map { case ((parent: String, child: String), callCount: Long) =>
        Dependencies(0, 0, Seq(DependencyLink(parent = parent, child = child, callCount = callCount)))
      }
      .reduce(_ + _) // merge DLs under one Dependencies object, which overrides +
```

The job also doesn't filter traces by timestamps, it just reads everything from the table, so a single record already presents fully aggregated result. If we want that job to generate data for adjacent time intervals, and considering that it clamps he timestamp to 24hrs, maybe it should be doing a filter on the traces. 

```
startTs = (ts floor 24hrs) - 24hr
endTs = ts floor 24hrs
```

**Comment by codefromthecrypt** [comment]
> I would prefer if the UI could simply display the time interval for which
> the aggregated results are shown.
> That

Adding multiple records together doesn't sound right to me - what if their
startTs-endTs intervals overlap or are disjoint? The aggregate numbers
wouldn't really make sense. At least if we're showing a single aggregate
record we can guarantee that the counts are consistent.

This sounds like edge case fishing. For example, this feature was
documented as allowing backends to store a different granularity. Even to
the point of explaining what that meant, ex. How we expect to treat a look
back that ends in-between grains. For example summing 2 days, if look back
is 36hrs. This issue doesn't seem like a feature change to me, but your
response to it does.

> Also, summing may work for counts, but once we start adding more
> measures, like p99, we won't be able to aggregate across multiple time
> intervals (maybe we'll need to go back to 'moments').
> This is really getting past the current feature. We have discussed that we
> might add summable measures such as bucketed latency intervals. Average or
> p99 is a distracting addition to this issue.
> 
> Incidentally, the current Spark job impl leaves startTs/endTs==0, so the
> maxBY above won't quite work.
> 
> ```
> val dependencies: Dependencies = aggregates
>   .map { case ((parent: String, child: String), callCount: Long) =>
>     Dependencies(0, 0, Seq(DependencyLink(parent = parent, child =
> ```
> 
> child, callCount = callCount)))
>       }
>       .reduce(_ + _) // merge DLs under one Dependencies object, which
> overrides +
> 
> The job also doesn't filter traces by timestamps, it just reads
> everything from the table, so a single record already presents fully
> aggregated result. If we want that job to generate data for adjacent time
> intervals, and considering that it clamps he timestamp to 24hrs, maybe it
> should be doing a filter on the traces.
> 
> startTs = (ts floor 24hrs) - 24hr
> endTs = ts floor 24hrs

This response is hard because it is tail wagging the dog. The spark job is
one implementation of the dependency api, and as far as I remember doesn't
even run the tests.

Let's focus on the feature we have today, fill missing tests and fix jobs
that implement the feature. If we want a new or different feature that
tries to sum P99, that's cool, but seems distracting to do that in the
middle of a big report.

In other words can we please put this train back on the rails.

I am willing to help fix the bug, but not if that means reworking zipkin's
code around a ramble of new requirements and UI changes.

We have 2 bugs: here we have a bug which we should be explainable in the
javadoc on DependencyStore and a DependencyStoreSpec

The other bug is that the spark job has no tests. This leads to
hypothetical discussion around features which could be avoided if it ever
actually ran our interop test. We shouldn't kite any more feature work
based on the assumptions baked into untested code. So, that's bug 2.

Once we fix the bugs I think we should start thinking about future features
but honestly the Cassandra related codebase has some serious tech debt to
pay off.

**Comment by yurishkuro** [comment]
If it's already established elsewhere that summing up multiple records is the expected behavior, then this would fix CassandraDependencyStore: https://github.com/openzipkin/zipkin/pull/869

``` scala
    FutureUtil.toFuture(repository.getDependencies(startEpochDayMillis, endEpochDayMillis))
      .map { dependencies =>
        dependencies.asScala
          .map(codec.decode(_))
          .map(thriftToDependencies(_).toDependencies)
          .flatMap(_.links)
          .groupBy(link => (link.parent, link.child))
          .map {
            case ((parent, child), links) => DependencyLink(parent, child, links.map(_.callCount).sum)
          }.toSeq
      }
```

**Comment by yurishkuro** [comment]
But the Spark job would produce duplicate results until this is fixed: https://github.com/openzipkin/zipkin-dependencies-spark/issues/7
